pydantic/_internal/_typing_extra.py

```diff
@@ -1,192 +1,414 @@
-    """Check if a given type is a `typing.Final` type."""
-    if v is None:
+# TODO implement `is_finalvar_annotation` as Final can be wrapped with other special forms:
+def is_finalvar(tp: Any, /) -> bool:
+    """Return whether the provided argument is a `Final` special form, parametrized or not.
+
+    ```python test="skip" lint="skip"
+    is_finalvar(Final[int])
+    #> True
+    is_finalvar(Final)
+    #> True
+    """
+    # Final is not necessarily parametrized:
+    return _is_typing_name(tp, name='Final') or _is_typing_name(get_origin(tp), name='Final')
+
+
+def is_required(tp: Any, /) -> bool:
+    """Return whether the provided argument is a `Required` special form.
+
+    ```python test="skip" lint="skip"
+    is_required(Required[int])
+    #> True
+    """
+    return _is_typing_name(get_origin(tp), name='Required')
+
+
+def is_not_required(tp: Any, /) -> bool:
+    """Return whether the provided argument is a `NotRequired` special form.
+
+    ```python test="skip" lint="skip"
+    is_required(Required[int])
+    #> True
+    """
+    return _is_typing_name(get_origin(tp), name='NotRequired')
+
+
+def is_no_return(tp: Any, /) -> bool:
+    """Return whether the provided argument is the `NoReturn` special form.
+
+    ```python test="skip" lint="skip"
+    is_no_return(NoReturn)
+    #> True
+    ```
+    """
+    return _is_typing_name(tp, name='NoReturn')
+
+
+def is_never(tp: Any, /) -> bool:
+    """Return whether the provided argument is the `Never` special form.
+
+    ```python test="skip" lint="skip"
+    is_never(Never)
+    #> True
+    ```
+    """
+    return _is_typing_name(tp, name='Never')
+
+
+_DEPRECATED_TYPES: tuple[type[typing_extensions.deprecated], ...] = (typing_extensions.deprecated,)
+if hasattr(warnings, 'deprecated'):
+    _DEPRECATED_TYPES = (*_DEPRECATED_TYPES, warnings.deprecated)  # pyright: ignore[reportAttributeAccessIssue]
+
+
+def is_deprecated_instance(obj: Any, /) -> TypeIs[deprecated]:
+    """Return whether the argument is an instance of the `warnings.deprecated` class or the `typing_extensions` backport."""
+    return isinstance(obj, _DEPRECATED_TYPES)
+
+
+_NONE_TYPES: tuple[Any, ...] = (None, NoneType, typing.Literal[None], typing_extensions.Literal[None])
+
+
+def is_none_type(tp: Any, /) -> bool:
+    """Return whether the argument represents the `None` type as part of an annotation.
+
+    ```python test="skip" lint="skip"
+    is_none_type(None)
+    #> True
+    is_none_type(NoneType)
+    #> True
+    is_none_type(Literal[None])
+    #> True
+    is_none_type(type[None])
+    #> False
+    """
+    return tp in _NONE_TYPES
+
+
+def is_namedtuple(tp: Any, /) -> bool:
+    """Return whether the provided argument is a named tuple class.
+
+    The class can be created using `typing.NamedTuple` or `collections.namedtuple`.
+    Parametrized generic classes are *not* assumed to be named tuples.
+    """
+    from ._utils import lenient_issubclass
```

Comment:
Should we just import this at the top level?

Reply:
circular import, added a comment